### ISSUE TITLE
FTPS uploads using TLSv1.3 are likely to fail unexpectedly

### ISSUE THREAD

**Issue by Castaglia** [proftpd/proftpd#959]
Building ProFTPD with `mod_tls`, using OpenSSL-1.1.1 or later, in order to support TLSv1.3, can have unexpected consequences.  Namely that FTPS uploads which use TLSv1.3 are likely to fail with:
```
2020-04-04 19:31:19,259 [11130] <netio:19>: using tls read() for data reading stream
2020-04-04 19:31:19,259 mod_tls/2.9[11130]: [msg] received protocol record message (content type = ApplicationData, len = 16401)
2020-04-04 19:31:19,259 [11130] <tls:17>: SSL_ERROR_SYSCALL error (errcode 0) occurred on line 8928; ignoring ECONNRESET (Connection reset by peer)
2020-04-04 19:31:19,259 [11130] <data:9>: aborting data transfer (errno = Connection reset by peer (104), quiet = false, true abort = false)
2020-04-04 19:31:19,259 [11130] <netio:19>: using tls shutdown() for data writing stream
```
And the client will see this error response:
```
450 Transfer aborted. Link to file server lost
```

Interestingly, this does not currently appear to happen for FTPS _downloads_ (or directory listings), only _uploads_.

The size of the data being uploaded seems to influence the likelihood of transfer failure.  Smaller sizes are more likely to succeed; larger sizes are more likely to fail.

Known workarounds so far: disable TLSv1.3 support:
```
<IfModule mod_tls.c>
  TLSEngine on
  ...
  TLSProtocol ALL -TLSv1.3
</IfModule>
```
which is not a desirable workaround, but it _does_ prevent user-visible failed transfers.

**Comment by Castaglia** [comment]
In researching this issue further, I found two relevant OpenSSL bugs/tickets:
* https://github.com/openssl/openssl/issues/7948
* https://github.com/openssl/openssl/issues/10880

The scenarios described in the above tickets match the FTPS data transfer use cases:
* For downloads (and directory listings), `mod_tls` only _writes_ data on the data TCP connection, not _read_
* For uploads, `mod_tls` only _reads_ data from the data TCP connection, not _write

**Comment by Castaglia** [comment]
For reproducing these upload issues, I've used `curl` with an OpenSSL backend:
```
$ dd if=/dev/urandom of=/tmp/file250MB.dat count=250 bs=1048576
$ curl -kvs -T /tmp/file250MB.dat --user ... --ftp-pasv --ssl ftp://proftpd:2121
...
* We are completely uploaded and fine
* Remembering we are in dir ""
} [5 bytes data]
* TLSv1.3 (OUT), TLS alert, close notify (256):
} [2 bytes data]
< 450 Transfer aborted. Link to file server lost
* server did not report OK, got 450
```

I attempted to see if this was an OpenSSL-specific issue by using a different FTPS client:
```
$ lftp -v
LFTP | Version 4.8.4 | Copyright (c) 1996-2017 Alexander V. Lukyanov
...
Libraries used: GnuTLS 3.6.11, idn2 2.2.0, Readline 8.0, zlib 1.2.11
```
For this GnuTLS-using FTPS client, uploads succeed regardless of the size.  This helps demonstrate that the issue might indeed be related to OpenSSL's implementation.

**Comment by Castaglia** [comment]
This fix cannot be backported to 1.3.6, since TLSv1.3 support _in general_ requires a lot more work that has not been done in the 1.3.6 branch.

**Comment by Castaglia** [comment]
I changed my mind, and decided that this was worth backporting to the 1.3.6 branch.
